compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -524,10 +524,13 @@ private BType setLiteralValueAndGetType(BLangLiteral literalExpr, BType expType)
                 }
             } else if (expType.tag == TypeTags.UNION) {
                 Set<BType> memberTypes = ((BUnionType) expType).getMemberTypes();
-                if (memberTypes.stream()
-                        .anyMatch(memType -> memType.tag == TypeTags.INT || memType.tag == TypeTags.JSON ||
-                                memType.tag == TypeTags.ANYDATA || memType.tag == TypeTags.ANY)) {
-                    return setLiteralValueAndGetType(literalExpr, symTable.intType);
+                for (BType memType : memberTypes) {
```

Comment:
I think we need to take all the members in the union into consideration before taking a desition. 
`any|byte b = 1` here `b` is `int`.